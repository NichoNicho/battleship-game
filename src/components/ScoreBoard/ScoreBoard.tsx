import React from "react";
import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper,
} from "@mui/material";
import { useTranslation } from "react-i18next";

type ShipStatus = {
  name: string;
  size: number;
  hits: number;
  sunk: boolean;
};

type ScoreBoardProps = {
  ships: ShipStatus[];
};

const ScoreBoard: React.FC<ScoreBoardProps> = ({ ships }) => {
  const { t } = useTranslation("playLocal");

  return (
    <TableContainer component={Paper} style={{ margin: "16px 0" }}>
      <Table size="small">
        <TableHead>
          <TableRow>
            <TableCell>{t("ship")}</TableCell>
            <TableCell align="center">{t("size")}</TableCell>
            <TableCell align="center">{t("hits")}</TableCell>
            <TableCell align="center">{t("status")}</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {ships.map(({ name, size, hits, sunk }) => (
            <TableRow key={name}>
              <TableCell>{name}</TableCell>
              <TableCell align="center">{size}</TableCell>
              <TableCell align="center">{hits}</TableCell>
              <TableCell align="center">
                {sunk ? t("sunk") : t("afloat")}
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default ScoreBoard;
