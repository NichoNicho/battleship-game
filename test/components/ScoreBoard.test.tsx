import { render, screen } from "@testing-library/react";
import ScoreBoard from "$components/ScoreBoard";
import "@testing-library/jest-dom";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (key: string) => key,
  }),
}));

const mockShips = [
  { name: "Battleship", size: 4, hits: 2, sunk: false },
  { name: "Destroyer", size: 3, hits: 3, sunk: true },
];

describe("ScoreBoard", () => {
  it("renders correctly", () => {
    const { container } = render(<ScoreBoard ships={mockShips} />);
    expect(container).toMatchSnapshot();
  });

  it("displays ship information correctly", () => {
    render(<ScoreBoard ships={mockShips} />);

    // Locate rows by ship names
    const destroyerRow = screen.getByText("Destroyer").closest("tr");
    const battleshipRow = screen.getByText("Battleship").closest("tr");

    // Validate Destroyer row
    expect(destroyerRow).toBeInTheDocument();
    expect(screen.getByText("Destroyer")).toBeInTheDocument();
    expect(screen.getAllByText("3").length).toBeGreaterThanOrEqual(2); // Ensures both size and hits are present
    expect(screen.getByText("sunk")).toBeInTheDocument();

    // Validate Battleship row
    expect(battleshipRow).toBeInTheDocument();
    expect(screen.getByText("Battleship")).toBeInTheDocument();
    expect(screen.getByText("4")).toBeInTheDocument();
    expect(screen.getByText("2")).toBeInTheDocument();
    expect(screen.getByText("afloat")).toBeInTheDocument();
  });
});
