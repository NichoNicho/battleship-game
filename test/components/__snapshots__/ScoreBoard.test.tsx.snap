// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreBoard renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    style="margin: 16px 0px;"
  >
    <table
      class="MuiTable-root css-rqglhn-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
            scope="col"
          >
            ship
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-12q85cb-MuiTableCell-root"
            scope="col"
          >
            size
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-12q85cb-MuiTableCell-root"
            scope="col"
          >
            hits
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-12q85cb-MuiTableCell-root"
            scope="col"
          >
            status
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Battleship
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
          >
            4
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
          >
            2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
          >
            afloat
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Destroyer
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
          >
            3
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
          >
            3
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
          >
            sunk
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
